### PR TITLE
feat(specs): Dencun P2P & Engine API changes

### PR DESCRIPTION
**Pull request by danyalprout**
**Description**

Specify the engine API and p2p changes for Dencun. See previous discussion in https://github.com/ethereum-optimism/optimism/pull/8840

### CODE REVIEW

File: specs/derivation.md [hunk 1]
```diff
@@ -798,24 +798,44 @@ To interact with the engine, the [execution engine API][exec-engine] is used, wi
...
+- Bedrock
+  - The withdrawals field MUST be nil
+  - The blob gas used field MUST be nil
+  - The blob gas limit field MUST be nil
+- Canyon
+  - The withdrawals field MUST be non-nil
+  - The withdrawals field MUST be an empty list
+  - The blob gas used field MUST be nil
+  - The blob gas limit field MUST be nil
+- Ecotone
```
Comment (protolambda) [inline specs/derivation.md:847]:
```suggestion
  - The blob gas limit field MUST be nil
- Delta: exactly like Canyon
- Ecotone
```

File: specs/derivation.md [hunk 2]
```diff
@@ -798,24 +798,44 @@ To interact with the engine, the [execution engine API][exec-engine] is used, wi
...
 [exec-engine]: exec-engine.md
 
-- [`engine_forkchoiceUpdatedV2`] — updates the forkchoice (i.e. the chain head) to `headBlockHash` if different, and
+- [`engine_forkchoiceUpdatedV3`] — updates the forkchoice (i.e. the chain head) to `headBlockHash` if different, and
   instructs the engine to start building an execution payload if the payload attributes parameter is not `null`.
-- [`engine_getPayloadV2`] — retrieves a previously requested execution payload build.
-- [`engine_newPayloadV2`] — executes an execution payload to create a block.
+- [`engine_getPayloadV3`] — retrieves a previously requested execution payload build.
+- `engine_newPayload`
+  - [`engine_newPayloadV2`] — executes a Bedrock or Canyon execution payload to create a block.
```
Comment (protolambda) [inline specs/derivation.md]:
```suggestion
  - [`engine_newPayloadV2`] — executes a Bedrock/Canyon/Delta execution payload to create a block.
```

File: specs/derivation.md [hunk 3]
```diff
@@ -907,8 +927,9 @@ To process unsafe payloads, the payload must:
 
 The payload is then processed with a sequence of:
 
-- `engine_newPayloadV2`: process the payload. It does not become canonical yet.
-- `engine_forkchoiceUpdatedV2`: make the payload the canonical unsafe L2 head, and keep the safe/finalized L2 heads.
+- `engine_newPayloadV3`: process the payload. It does not become canonical yet. (Bedrock and Canyon
+payloads MUST use V2)
```
Comment (protolambda) [inline specs/derivation.md]: If we enforce V2 usage for pre-Ecotone, then we should keep V2 around in the specification, and not overwrite it with V3 specs.
Also, we generally want to keep pre-Ecotone specs around in a compatible way, until Ecotone has been accepted by governance.

**Review by protolambda — CHANGES_REQUESTED** [review]
The changes look good, but we should be extending, not overwriting, the specs.

File: specs/overview.md [hunk 4]
```diff
@@ -185,14 +185,17 @@ For each iteration of the block derivation loop described above, the rollup driv
...
-includes a block hash as one of its fields.
-3. Call `engine_newPayloadV2` with the payload returned in step 2.
-4. Call `engine_forkchoiceUpdatedV2` with the fork choice parameter's `headBlockHash` set to the block hash returned in
-step 2. The tip of the L2 chain is now the block created in step 1.
+1. Call [fork choice updated][EngineAPIVersion] with the payload attributes object. We'll skip over the details of the
+fork choice state parameter for now - just know that one of its fields is the L2 chain's `headBlockHash`, and that it
+is set to the block hash of the tip of the L2 chain. The Engine API returns a payload ID.
+2. Call [get payload][EngineAPIVersion] with the payload ID returned in step 1. The engine API returns a payload object
+that includes a block hash as one of its fields.
+3. Call [new payload][EngineAPIVersion] with the payload returned in step 2. (Pre-Ecotone blocks MUST use the V2
```
Comment (EvanJRichard) [inline specs/overview.md]: Consider describing what happens to Post-Ecotone blocks. (They `MUST` use V3, right?) I'd accept an argument that it's unnecessary extra info, though.

File: specs/rollup-node-p2p.md [hunk 5]
```diff
@@ -258,28 +259,37 @@ Pre-Canyon/Shanghai blocks are broadcast on `/optimism/<chainId>/0/blocks`.
 
 Post-Canyon/Shanghai blocks are broadcast on `/optimism/<chainId>/1/blocks`.
```
Comment (EvanJRichard) [inline specs/rollup-node-p2p.md]: Style question: do we want to change this to `Post-Canyon/Shanghai, but Pre-Ecotone`? Another "clarity versus over-verbosity" balancing act I'm wondering about.

**Review by EvanJRichard — APPROVED** [review]
LGTM, I especially like the deduplication through linking to versions.

File: specs/rollup-node-p2p.md (hunk 5 above)
Comment (danyalprout) [reply specs/rollup-node-p2p.md]: I think that's a good call, let me update these topics to list which hardforks broadcast on them

File: specs/derivation.md [hunk 6]
```diff
@@ -798,24 +800,56 @@ To interact with the engine, the [execution engine API][exec-engine] is used, wi
...
+
+[eth-payload]: https://github.com/ethereum/execution-apis/blob/main/src/engine/cancun.md
+
+The `ExecutionPayload` has the following requirements:
+
+- Bedrock
+  - The withdrawals field MUST be nil
+  - The blob gas used field MUST be nil
+  - The blob gas limit field MUST be nil
+- Canyon
```
Comment (tynes) [inline specs/derivation.md]: nit: `Canyon, Delta` instead of another line that says `Delta: exactly like Canyon`

File: specs/rollup-node-p2p.md [hunk 7]
```diff
@@ -290,8 +300,13 @@ An [extended-validator] checks the incoming messages as follows, in order of ope
 - `[REJECT]` if the `payload.timestamp` is more than 5 seconds into the future
 - `[REJECT]` if the `block_hash` in the `payload` is not valid
 - `[REJECT]` if the block is on the V1 topic and has withdrawals
-- `[REJECT]` if the block is on the V2 topic and does not have withdrawals
-- `[REJECT]` if the block is on the V2 topic and has a non-zero amount of withdrawals
+- `[REJECT]` if the block is on a topic >= V2 and does not have withdrawals
+- `[REJECT]` if the block is on a topic >= V2 and has a non-zero amount of withdrawals
```
Comment (protolambda) [inline specs/rollup-node-p2p.md]:
```suggestion
- `[REJECT]` if the block is on a topic <= V2 and has a withdrawals list
- `[REJECT]` if the block is on a topic >= V2 and does not have an empty withdrawals list
```

File: specs/rollup-node-p2p.md [hunk 8]
```diff
@@ -290,8 +300,13 @@ An [extended-validator] checks the incoming messages as follows, in order of ope
...
+- `[REJECT]` if the block is on a topic <= V2 and has a blob gas value set
```
Comment (protolambda) [inline specs/rollup-node-p2p.md]:
```suggestion
- `[REJECT]` if the block is on a topic <= V2 and has a blob gas-used value set
```

File: specs/rollup-node-p2p.md [hunk 9]
```diff
@@ -290,8 +300,13 @@ An [extended-validator] checks the incoming messages as follows, in order of ope
...
+- `[REJECT]` if the block is on a topic <= V2 and has an excess blob gas value set
+- `[REJECT]` if the block is on a topic >= V3 and has a blob gas used value that is not zero
```
Comment (protolambda) [inline specs/rollup-node-p2p.md]:
```suggestion
- `[REJECT]` if the block is on a topic >= V3 and has a blob gas-used value that is not zero
```

File: specs/rollup-node-p2p.md [hunk 10]
```diff
@@ -290,8 +300,13 @@ An [extended-validator] checks the incoming messages as follows, in order of ope
...
+- `[REJECT]` if the block is on a topic >= V3 and has an excess blob gas value that is not zero
+- `[REJECT]` if the block is on a topic >= V3 and the parent beacon block root is nil
```
Comment (protolambda) [inline specs/rollup-node-p2p.md:310]:
```suggestion
- `[REJECT]` if the block is on a topic <= V2 and the parent beacon block root is not nil
- `[REJECT]` if the block is on a topic >= V3 and the parent beacon block root is nil
```

**Review by protolambda — APPROVED** [review]
LGTM, just a few nits on the gossip validation wording, and a missing case on  the parent-beacon-block-root check. It's already implied by the topic encoding/decoding, but I think it makes sense to add, since we validate the other attribute changes the same way.

File: specs/rollup-node-p2p.md (hunk 7 above)
Comment (danyalprout) [reply specs/rollup-node-p2p.md]: For this one I believe:
> - `[REJECT]` if the block is on a topic <= V2 and has a withdrawals list

Should be:
> - `[REJECT]` if the block is on a topic < V2 and has a withdrawals list

As the V2 topic has withdrawals. I'll update to this.

Comment (danyalprout) [reply specs/rollup-node-p2p.md]: Done in https://github.com/ethereum-optimism/optimism/pull/8852/commits/8d44f673c2de435079885fb189294f7dd76e6c2d

**Review by protolambda — APPROVED** [review]

**Comment by tynes** [comment]
This needs a rebase to get `kontrol-tests` to pass, cc @danyalprout

**Review by tynes — APPROVED** [review]

**Pull request merged**